### ISSUE TITLE
fail safe bootloader fails in upgrading image

### ISSUE THREAD

**Issue by AnandSingh** [Microchip-MPLAB-Harmony/bootloader_apps_uart#1]
I am using [SAM E54](https://www.microchip.com/developmenttools/productdetails/atsame54-xpro) to verify the fail safe [uart bootlaoder](https://github.com/Microchip-MPLAB-Harmony/bootloader_apps_uart/tree/master/apps/uart_fail_safe_bootloader)

I followed these [instruction](https://github.com/Microchip-MPLAB-Harmony/bootloader_apps_uart/blob/master/apps/uart_fail_safe_bootloader/docs/readme_sam_e54_xpro.md)

Below the issue which I am facing ?
1. While updating the other bank with the merged binary, its failing in verification steps
```
python3 btl_host.py -v -s -i /dev/ttyACM0 -d same5x -a 0x80000 -f btl_app_merged.bin

Unlocking

Programming: |||||||||||||||||||||||||||||||||||||||||||||||||||| 100.0% Complete 

Verification

Error: ... fail (status = 0x54)
```

2. Also sometime when doing the merged binary update, it throw below error message 
```
python3 btl_host.py -v -s -i /dev/ttyACM0 -d same5x -a 0x80000 -f btl_app_merged.bin 

Unlocking


Error: invalid response code (0x51). Check that your file size and address are correct.
```

So basically was not able to proceed with any upgrade . Can you please confirm at your end having the same issue.

**Comment by amitraddi** [comment]
Hi Anand,

Thanks for posting the issue.
Seeing the com port you are using /dev/ttyACM0 i assume you are trying to run demo on linux machine. I have not tried to run on Linux machine though but the demo by following the instructions provided works on windows machine.

1. When generating the bootloader binary for merging with the test_app binary, please make sure you have disabled the fuse settings for bootloader project and **regenerated** (mentioned in the instructions).
2. Make sure the offset passed during merging of the binary is 0x2000
   python <harmony3_path>\bootloader\tools\btl_app_merge_bin.py -o **0x2000** -b <harmony3_path>\bootloader_apps_uart\apps\uart_fail_safe_bootloader\bootloader\firmware\sam_e54_xpro.X\dist\sam_e54_xpro\production\sam_e54_xpro.X.production.bin -a <harmony3_path>\bootloader_apps_uart\apps\uart_fail_safe_bootloader\test_app\firmware\sam_e54_xpro.X\dist\sam_e54_xpro\production\sam_e54_xpro.X.production.bin

**Comment by AnandSingh** [comment]
Thanks Amit for your reply. I did tried these option as mentioned by you. It's the same result.

**Comment by amitraddi** [comment]
Hi @AnandSingh 

Failure for verification can have these two possible reasons
1. There was an issue in Programming operation (Which i do not suspect)
2. The bytes received over UART are not the same as the actual binary. either there has been a data loss or some junk characters have been received.

Can you see if there is any Overflow over the UART while receiving the data. To do that you can add below piece of code in input_task() function

```
    USART_ERROR errorStatus = USART_ERROR_NONE;

    errorStatus = SERCOM2_USART_ErrorGet();

    if (errorStatus != USART_ERROR_NONE)
    {
        // You should get the error code printed on the terminal
        SERCOM2_USART_WriteByte(errorStatus);
        return;
    }
```

![image](https://user-images.githubusercontent.com/45988337/90717813-19ab0480-e2ce-11ea-9ce7-0a49a598bb10.png)

**Comment by AnandSingh** [comment]
Sure Le me try this option and will update you accordingly.

On another note I also tried the [bootloader_apps_ethernet](https://github.com/Microchip-MPLAB-Harmony/bootloader_apps_ethernet) And it worked perfect. I have two question 
1. Can I implement the fail safe bootloader design to bootloader_apps_ethenet, i.e. upgrading to the in-active bank and do a bank switch after merged binary upgrade.
2. Where can I find the implementation (source code) of the [UnifiedHost-1.16.2.jar](https://github.com/Microchip-MPLAB-Harmony/bootloader/tree/master/tools/UnifiedHost-1.16.2)

**Comment by amitraddi** [comment]
@AnandSingh 

Sure please try to get the UART Fail safe Bootloader running. as this is optimized saves lot of space.

Regarding to UDP Fail safe bootloader. yes it can be modified to program and run from Inactive bank. I did try this out and i was able to get it up and running. Attaching bot the bootloader and test application for the same 

[udp_fail_safe_bootloader.zip](https://github.com/Microchip-MPLAB-Harmony/bootloader_apps_uart/files/5101561/udp_fail_safe_bootloader.zip)

1. Existing bootloader files have to be modified to always program in Inactive bank (0x80000). below are the changes done

 **bootloader_nvm_interface.h**
![image](https://user-images.githubusercontent.com/45988337/90730833-6a2e5c00-e2e6-11ea-8310-38be4abcc852.png)

**bootloader_nvm_interface.c**
![image](https://user-images.githubusercontent.com/45988337/90730949-9ba72780-e2e6-11ea-9049-86d47f5ee88b.png)

![image](https://user-images.githubusercontent.com/45988337/90731109-de68ff80-e2e6-11ea-9f4a-84e6b0ec5d25.png)

2. Perform Bank Swap and Reset command instead of NVIC_Reset

![image](https://user-images.githubusercontent.com/45988337/90731229-0bb5ad80-e2e7-11ea-8fb2-546eb5da4998.png)

3. Build and Program the bootloader project

4. Disable Fuse settings for Bootloader project and generate code so that it can be updated into other bank. **Make sure you not to overwrite the changes done to bootloader code**

![image](https://user-images.githubusercontent.com/45988337/90731427-50414900-e2e7-11ea-9f7f-c5feb0d50ae8.png)

5. Add the Bootloader project as loadable project for in test_app to generate a combined unified hex file to be programmed at **0x80000** 

![image](https://user-images.githubusercontent.com/45988337/90731615-867ec880-e2e7-11ea-8d95-a610d9525edc.png)

6. Build the test_app project. it will build both bootloader and test app and Once done **sam_e54_xpro.X.production.unified.hex** will be generated.

7. Use this unified hex file with UnifiedHost-1.16.2.jar to program.

![image](https://user-images.githubusercontent.com/45988337/90732136-48ce6f80-e2e8-11ea-9139-c404fbf41fec.png)

8. Output after programming the hex file twice

![image](https://user-images.githubusercontent.com/45988337/90731828-d493cc00-e2e7-11ea-8571-7ad1f5f2f259.png)

**Note: With this approach you can only program the combined hex file of test application and bootloader**

*Where can I find the implementation (source code) of the UnifiedHost-1.16.2.jar*

Currently the source code is not public. let me get back to you on this. but with above changes you may not need any changes to the host app

**Comment by AnandSingh** [comment]
Thanks a lot, this saves lot of time. Eventually we will be using the
Ethernet for upgrading firmware. This is the best start for us.
Once again thanks and appreciate your support. Will check at my end.

**Comment by AnandSingh** [comment]
I verified these changes and it work great.

**Comment by amitraddi** [comment]
Thats good to know.
